File: web-frontend/src/types/dataImport.ts
export type ImportJobStatus = 'pending' | 'processing' | 'completed' | 'failed' | 'cancelled'
export type ImportType = 'trading_data' | 'station_output' | 'storage_operation'
export type EmsFormat = 'standard' | 'sungrow' | 'huawei' | 'catl'
export type AnomalyType = 'missing' | 'format_error' | 'out_of_range' | 'duplicate'
export type AnomalyStatus = 'pending' | 'confirmed_normal' | 'corrected' | 'deleted'

export const IMPORT_TYPE_LABELS: Record<ImportType, string> = {
  trading_data: '交易数据',
  station_output: '电站出力数据',
  storage_operation: '储能运行数据',
}

export interface ImportJob {
  id: string
  file_name: string
  original_file_name: string
  file_size: number
  station_id: string
  import_type: ImportType
  status: ImportJobStatus
  total_records: number
  processed_records: number
  success_records: number
  failed_records: number
  data_completeness: number
  last_processed_row: number
  celery_task_id: string | null
  error_message: string | null
  started_at: string | null
  completed_at: string | null
  imported_by: string
  created_at: string
  updated_at: string
}

export interface ImportJobListResponse {
  items: ImportJob[]
  total: number
  page: number
  page_size: number
}

export interface ImportAnomaly {
  id: string
  import_job_id: string
  row_number: number
  anomaly_type: AnomalyType
  field_name: string
  raw_value: string | null
  description: string
  status: AnomalyStatus
  created_at: string
  updated_at: string
}

export interface AnomalySummary {
  anomaly_type: string
  count: number
}

export interface ImportResult {
  total_records: number
  success_records: number
  failed_records: number
  data_completeness: number
  anomaly_summary: AnomalySummary[]
}

export interface ImportAnomalyListResponse {
  items: ImportAnomaly[]
  total: number
  page: number
  page_size: number
}

// --- 异常管理 (Story 2.4) ---

export interface AnomalyCorrectRequest {
  corrected_value: string
}

export interface AnomalyBulkActionRequest {
  anomaly_ids: string[]
  action: 'delete' | 'confirm_normal'
}

export interface AnomalyBulkActionResponse {
  affected_count: number
  action: string
}

export interface AnomalyDetail extends ImportAnomaly {
  original_file_name?: string | null
  station_id?: string | null
}

export interface AnomalyStatsSummaryItem {
  anomaly_type: string
  count: number
}

// --- 电站出力与储能运行数据 (Story 2.5) ---

export interface StationOutputRecord {
  id: string
  trading_date: string
  period: number
  station_id: string
  actual_output_kw: number
  import_job_id: string
  created_at: string
}

export interface StationOutputRecordListResponse {
  items: StationOutputRecord[]
  total: number
  page: number
  page_size: number
}

export interface StorageOperationRecord {
  id: string
  trading_date: string
  period: number
  device_id: string
  soc: number
  charge_power_kw: number
  discharge_power_kw: number
  cycle_count: number
  import_job_id: string
  created_at: string
}

export interface StorageOperationRecordListResponse {
  items: StorageOperationRecord[]
  total: number
  page: number
  page_size: number
}

File: api-server/app/schemas/data_import.py
from datetime import date, datetime
from decimal import Decimal
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, field_validator

ImportJobStatus = Literal["pending", "processing", "completed", "failed", "cancelled"]
ImportType = Literal["trading_data", "station_output", "storage_operation", "market_data"]
EmsFormat = Literal["standard", "sungrow", "huawei", "catl"]
AnomalyType = Literal["missing", "format_error", "out_of_range", "duplicate"]
AnomalyStatus = Literal["pending", "confirmed_normal", "corrected", "deleted"]


# --- 导入任务 schemas ---


class ImportJobCreate(BaseModel):
    station_id: UUID
    import_type: ImportType = "trading_data"
    ems_format: EmsFormat | None = None


class ImportJobRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    file_name: str
    original_file_name: str
    file_size: int
    station_id: UUID
    import_type: ImportType
    ems_format: EmsFormat | None
    status: ImportJobStatus
    total_records: int
    processed_records: int
    success_records: int
    failed_records: int
    data_completeness: Decimal
    last_processed_row: int
    celery_task_id: str | None
    error_message: str | None
    started_at: datetime | None
    completed_at: datetime | None
    imported_by: UUID
    created_at: datetime
    updated_at: datetime


class ImportJobListResponse(BaseModel):
    items: list[ImportJobRead]
    total: int
    page: int
    page_size: int


# --- 异常记录 schemas ---


class ImportAnomalyRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    import_job_id: UUID
    row_number: int
    anomaly_type: AnomalyType
    field_name: str
    raw_value: str | None
    description: str
    status: AnomalyStatus
    created_at: datetime
    updated_at: datetime


class ImportAnomalySummary(BaseModel):
    anomaly_type: str
    count: int


class ImportResultRead(BaseModel):
    total_records: int
    success_records: int
    failed_records: int
    data_completeness: Decimal
    anomaly_summary: list[ImportAnomalySummary]


class ImportAnomalyListResponse(BaseModel):
    items: list[ImportAnomalyRead]
    total: int
    page: int
    page_size: int


# --- 异常管理 schemas (Story 2.4) ---


class AnomalyCorrectRequest(BaseModel):
    corrected_value: str

    @field_validator("corrected_value")
    @classmethod
    def validate_non_empty(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("修正值不能为空")
        return v.strip()


class AnomalyBulkActionRequest(BaseModel):
    anomaly_ids: list[UUID]
    action: Literal["delete", "confirm_normal"]

    @field_validator("anomaly_ids")
    @classmethod
    def validate_non_empty(cls, v: list[UUID]) -> list[UUID]:
        if not v:
            raise ValueError("anomaly_ids 不能为空")
        if len(v) > 100:
            raise ValueError("单次批量操作最多 100 条")
        return v


class AnomalyBulkActionResponse(BaseModel):
    affected_count: int
    action: str


class AnomalyDetailRead(ImportAnomalyRead):
    """扩展 ImportAnomalyRead，关联 import_job 的文件名和电站信息。"""

    original_file_name: str | None = None
    station_id: UUID | None = None


# --- 电站出力数据 schemas ---


class StationOutputRecordRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    trading_date: date
    period: int
    station_id: UUID
    actual_output_kw: Decimal
    import_job_id: UUID
    created_at: datetime


class StationOutputRecordListResponse(BaseModel):
    items: list[StationOutputRecordRead]
    total: int
    page: int
    page_size: int


# --- 储能运行数据 schemas ---


class StorageOperationRecordRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    trading_date: date
    period: int
    device_id: UUID
    soc: Decimal
    charge_power_kw: Decimal
    discharge_power_kw: Decimal
    cycle_count: int
    import_job_id: UUID
    created_at: datetime


class StorageOperationRecordListResponse(BaseModel):
    items: list[StorageOperationRecordRead]
    total: int
    page: int
    page_size: int
